### PR TITLE
posm: Clear

### PR DESCRIPTION
**Pull request by saucepoint**
## Related Issue
Adds support for `poolManager.clear()` to posm

## Description of changes

Expose support for `Actions.CLEAR`, which requires a maximum allowable clear limit.

i.e. For cases where the user will not know the amount (autocompounding), a user may be _willing to forfiet up to 100 wei of dust_

### CODE REVIEW

File: src/PositionManager.sol [hunk 1]
```diff
@@ -161,6 +164,14 @@ contract PositionManager is
         }
     }
 
+    /// @dev integrators may elect to forfeit positive deltas with clear
+    /// provide a safety check that amount-to-clear is less than a user-provided maximum
+    function _clear(Currency currency, uint256 amountMax) internal {
+        int256 currencyDelta = poolManager.currencyDelta(address(this), currency);
+        if (amountMax < uint256(currencyDelta)) revert ClearExceedsMaxAmount(currency, currencyDelta, amountMax);
```
Comment (snreynolds) [inline src/PositionManager.sol:171]: this is preference but it reads easier to me if we flip the vars

if (currencyDelta > amountMax) revert..

File: test/libraries/CalldataDecoder.t.sol [hunk 2]
```diff
@@ -72,6 +72,14 @@ contract CalldataDecoderTest is Test {
         assertEq(Currency.unwrap(currency), Currency.unwrap(_currency));
     }
 
+    function test_fuzz_decodeCurrencyAndAddress(Currency _currency, uint256 _amount) public view {
```
Comment (snreynolds) [inline test/libraries/CalldataDecoder.t.sol:75]:
```suggestion
    function test_fuzz_decodeCurrencyAndUint256(Currency _currency, uint256 _amount) public view {
```

File: test/position-managers/IncreaseLiquidity.t.sol [hunk 3]
```diff
@@ -458,4 +464,40 @@ contract IncreaseLiquidityTest is Test, PosmTestSetup, Fuzzers {
         assertEq(currency0.balanceOf(address(this)), balanceBefore0 - amount0);
         assertEq(currency1.balanceOf(address(this)), balanceBefore1 - amount1);
     }
+
+    function test_increaseLiquidity_clearRevert() public {
```
Comment (snreynolds) [inline test/position-managers/IncreaseLiquidity.t.sol:468]: can you add one test where someone tries to clear a negative balance, it should revert in core

File: test/position-managers/PositionManager.gas.t.sol [hunk 4]
```diff
@@ -170,13 +170,60 @@ contract PosMGasTest is Test, PosmTestSetup, GasSnapshot {
...
+        // Alice elects to forfeit unclaimed tokens
+        Plan memory planner = Planner.init();
+        planner.add(Actions.INCREASE_LIQUIDITY, abi.encode(tokenIdAlice, config, liquidityDelta, ZERO_BYTES));
+        planner.add(Actions.CLEAR, abi.encode(config.poolKey.currency0, halfTokensOwedAlice + 1 wei));
+        planner.add(Actions.CLEAR, abi.encode(config.poolKey.currency1, halfTokensOwedAlice + 1 wei));
+        bytes memory calls = planner.encode();
+
+        vm.prank(alice);
+        lpm.modifyLiquidities(calls, _deadline);
+        snapLastCall("PositionManager_increase_autocompound_clearExcess");
```
Comment (snreynolds) [inline test/position-managers/PositionManager.gas.t.sol:224]: can you keep the same gas test but with take, not clear

File: test/position-managers/IncreaseLiquidity.t.sol [hunk 5]
```diff
@@ -164,14 +165,19 @@ contract IncreaseLiquidityTest is Test, PosmTestSetup, Fuzzers {
...
-        vm.startPrank(alice);
-        increaseLiquidity(tokenIdAlice, config, liquidityDelta, ZERO_BYTES);
-        vm.stopPrank();
+        Plan memory planner = Planner.init();
+        planner.add(Actions.INCREASE_LIQUIDITY, abi.encode(tokenIdAlice, config, liquidityDelta, ZERO_BYTES));
+        planner.add(Actions.CLEAR, abi.encode(config.poolKey.currency0, 1 wei)); // alice is willing to forfeit 1 wei
+        planner.add(Actions.CLEAR, abi.encode(config.poolKey.currency1, 1 wei));
+        bytes memory calls = planner.encode();
 
-        // alice barely spent any tokens
```
Comment (snreynolds) [inline test/position-managers/IncreaseLiquidity.t.sol:171]: also can we leave this test with take, and just add the clear one as well

so we have:

test_increaseLiquidity_withExactFees_take
test_increaseLiquidity_withExactFees_clear

File: test/position-managers/PositionManager.gas.t.sol (hunk 4 above)
Comment (saucepoint) [reply test/position-managers/PositionManager.gas.t.sol:224]: we have one, where its using close (take)

`test_gas_autocompound_excessFeesCredit / PositionManager_increase_autocompoundExcessFeesCredit` (169,831)

**Review by snreynolds — APPROVED** [review]
lgtm but lets also get @hensha256 to review

File: src/libraries/CalldataDecoder.sol [hunk 6]
```diff
@@ -129,6 +129,17 @@ library CalldataDecoder {
         }
     }
 
+    function decodeCurrencyAndUint256(bytes calldata params)
```
Comment (hensha256) [inline src/libraries/CalldataDecoder.sol:132]:
```
 /// @dev equivalent to: abi.decode ...
```

File: src/PositionManager.sol [hunk 7]
```diff
@@ -204,6 +207,14 @@ contract PositionManager is
         }
     }
 
+    /// @dev integrators may elect to forfeit positive deltas with clear
+    /// provide a safety check that amount-to-clear is less than a user-provided maximum
```
Comment (hensha256) [inline src/PositionManager.sol:211]:
```suggestion
    /// provides a safety check that amount-to-clear is less than a user-provided maximum
```

File: src/PositionManager.sol [hunk 8]
```diff
@@ -204,6 +207,14 @@ contract PositionManager is
...
+    function _clear(Currency currency, uint256 amountMax) internal {
+        int256 currencyDelta = poolManager.currencyDelta(address(this), currency);
+        if (uint256(currencyDelta) > amountMax) revert ClearExceedsMaxAmount(currency, currencyDelta, amountMax);
```
Comment (hensha256) [inline src/PositionManager.sol:214]: more a stylistic thing - but I wonder if the behaviour should be that

```
if (uint256(currencyDelta) > amountMax) _take(currency, amount)
```

i feel like the natural behaviour is "if its less than X its cheaper for me to leave it, but if its more than X I want it" ?

Comment (saucepoint) [reply src/PositionManager.sol:214]: from offline: we'll introduce CLEAR_OR_TAKE as a command. we want to try to map commands ~1:1 with pool manager

**Review by hensha256 — APPROVED** [review]
